File: traineebox/internal/dialog/validator.go
package dialog

import (
	"encoding/json"
	"fmt"
	"strings"
)

// ScenarioSnapshot is the JSON canon owned by traineebox (AD-7).
// Dialog rejects `open` with 400 on drift: unknown slot, missing opening,
// dangling critical/requires. Mirrors dialog/validator.py: mode=both,
// Fact.audio, runtime meta and forbidden keys are rejected here so PUT
// never stores what open would refuse with 400 (spec E).
type ScenarioSnapshot struct {
	ID         string         `json:"id"`
	Opening    string         `json:"opening"`
	Mode       string         `json:"mode"`
	Critical   []string       `json:"critical"`
	Facts      []FactSnapshot `json:"facts"`
	Disclosure string         `json:"-"`
}

type FactSnapshot struct {
	Key        string            `json:"key"`
	Slot       string            `json:"slot"`
	Answers    map[string]string `json:"answers"`
	Requires   []string          `json:"requires,omitempty"`
	Disclosure string            `json:"disclosure,omitempty"`
}

// forbiddenTop mirrors Python FORBIDDEN_KEYS: never in a stored snapshot.
var forbiddenTop = []string{
	"aliases", "overrides", "since", "urge", "mood", "preset", "presets",
	"drift", "preview", "recording_path", "meta", "pin",
}

// Snapshot shape canon: unknown top-level keys are drift, not data (AD-7).
// Mirrored in dialog/validator.py — keep identical.
var allowedTop = map[string]bool{
	"id": true, "opening": true, "mode": true, "critical": true, "facts": true,
}

// Size caps mirrored in dialog/validator.py — keep identical.
const (
	MaxOpeningLen = 2000
	MaxFacts      = 256
)

// Validate checks snapshot against known slot ids.
// facts[].slot must be in dialog_slots; critical/requires must link existing facts.
func Validate(raw []byte, knownSlots map[string]bool) (ScenarioSnapshot, error) {
	var probe map[string]any
	if err := json.Unmarshal(raw, &probe); err != nil {
		return ScenarioSnapshot{}, fmt.Errorf("bad scenario json: %w", err)
	}
	for _, k := range forbiddenTop {
		if _, ok := probe[k]; ok {
			return ScenarioSnapshot{}, fmt.Errorf("forbidden key %q in snapshot", k)
		}
	}
	for k := range probe {
		if !allowedTop[k] {
			return ScenarioSnapshot{}, fmt.Errorf("unknown key %q in snapshot", k)
		}
	}
	if facts, ok := probe["facts"].([]any); ok {
		for _, item := range facts {
			if m, ok := item.(map[string]any); ok {
				if _, ok := m["audio"]; ok {
					return ScenarioSnapshot{}, fmt.Errorf("Fact.audio forbidden in snapshot")
				}
			}
		}
	}
	var sc ScenarioSnapshot
	if err := json.Unmarshal(raw, &sc); err != nil {
		return ScenarioSnapshot{}, fmt.Errorf("bad scenario json: %w", err)
	}
	if mode := sc.Mode; mode != "" && mode != "voice" && mode != "text" {
		if mode == "both" {
			return ScenarioSnapshot{}, fmt.Errorf("mode=both forbidden")
		}
		return ScenarioSnapshot{}, fmt.Errorf("bad mode %q", mode)
	}
	if sc.ID == "" {
		return ScenarioSnapshot{}, fmt.Errorf("scenario.id required")
	}
	if strings.TrimSpace(sc.Opening) == "" {
		return ScenarioSnapshot{}, fmt.Errorf("scenario.opening required")
	}
	if len([]rune(sc.Opening)) > MaxOpeningLen {
		return ScenarioSnapshot{}, fmt.Errorf("scenario.opening too long")
	}
	if len(sc.Facts) == 0 {
		return ScenarioSnapshot{}, fmt.Errorf("scenario.facts empty")
	}
	if len(sc.Facts) > MaxFacts {
		return ScenarioSnapshot{}, fmt.Errorf("scenario.facts too many")
	}
	keys := map[string]bool{}
	for _, f := range sc.Facts {
		if f.Key == "" || f.Slot == "" {
			return ScenarioSnapshot{}, fmt.Errorf("fact key/slot required")
		}
		if !knownSlots[f.Slot] {
			return ScenarioSnapshot{}, fmt.Errorf("unknown slot %q for fact %q", f.Slot, f.Key)
		}
		if len(f.Answers) == 0 || strings.TrimSpace(f.Answers["plain"]) == "" {
			return ScenarioSnapshot{}, fmt.Errorf("fact %q needs answers.plain", f.Key)
		}
		if d := f.Disclosure; d != "" && d != "volunteered" && d != "on_request" {
			return ScenarioSnapshot{}, fmt.Errorf("fact %q: bad disclosure %q", f.Key, d)
		}
		if keys[f.Key] {
			return ScenarioSnapshot{}, fmt.Errorf("duplicate fact %q", f.Key)
		}
		keys[f.Key] = true
	}
	for _, c := range sc.Critical {
		if !keys[c] {
			return ScenarioSnapshot{}, fmt.Errorf("critical %q not a fact", c)
		}
	}
	for _, f := range sc.Facts {
		for _, r := range f.Requires {
			if !keys[r] {
				return ScenarioSnapshot{}, fmt.Errorf("fact %q requires unknown %q", f.Key, r)
			}
		}
	}
	return sc, nil
}

File: dialog/validator.py
# -*- coding: utf-8 -*-
"""Snapshot JSON Schema canon (AD-7). Dialog rejects `open` with 400 on drift.

Forbidden in snapshots (spec Never): aliases/overrides/since/urge/mood/
presets/drift, partial-preview, recording_path, runtime meta, mode=both,
student-dials-PIN. The validator rejects them so the frozen dataclasses
(which still carry those fields for file-replay parity) never see them.
"""

from __future__ import annotations

from typing import Any

FORBIDDEN_KEYS = (
    "aliases", "overrides", "since", "urge", "mood", "preset", "presets",
    "drift", "preview", "recording_path", "meta", "pin", "mode",
)

# Snapshot shape canon: unknown top-level keys are drift, not data (AD-7).
# Mirrored in traineebox/internal/dialog/validator.go — keep identical.
ALLOWED_TOP_KEYS = ("id", "opening", "mode", "critical", "facts")
MAX_OPENING_LEN = 2000
MAX_FACTS = 256

DISCLOSURES = ("volunteered", "on_request")


class SnapshotError(ValueError):
    pass


def validate_snapshot(raw: dict[str, Any], known_slots: set[str] | None = None) -> dict[str, Any]:
    """Validate open snapshot. Returns normalized dict or raises SnapshotError."""
    if not isinstance(raw, dict):
        raise SnapshotError("scenario must be an object")
    for k in FORBIDDEN_KEYS:
        if k in raw and k not in ("mode",):
            raise SnapshotError(f"forbidden key {k!r} in snapshot")
    for k in raw:
        if k not in ALLOWED_TOP_KEYS:
            raise SnapshotError(f"unknown key {k!r} in snapshot")
    mode = raw.get("mode", "voice")
    if mode == "both":
        raise SnapshotError("mode=both forbidden")
    if mode not in ("voice", "text"):
        raise SnapshotError(f"bad mode {mode!r}")
    sid = raw.get("id") or ""
    if not sid:
        raise SnapshotError("scenario.id required")
    opening = raw.get("opening")
    if not isinstance(opening, str) or not opening.strip():
        raise SnapshotError("scenario.opening required")
    if len(opening) > MAX_OPENING_LEN:
        raise SnapshotError("scenario.opening too long")
    facts = raw.get("facts")
    if not isinstance(facts, list) or not facts:
        raise SnapshotError("scenario.facts empty")
    if len(facts) > MAX_FACTS:
        raise SnapshotError("scenario.facts too many")
    keys: set[str] = set()
    for f in facts:
        if not isinstance(f, dict):
            raise SnapshotError("fact must be an object")
        key, slot = f.get("key") or f.get("id") or "", f.get("slot") or ""
        if not key or not slot:
            raise SnapshotError("fact key/slot required")
        if known_slots is not None and slot not in known_slots:
            raise SnapshotError(f"unknown slot {slot!r} for fact {key!r}")
        answers = f.get("answers") or {}
        if not isinstance(answers, dict) or not (answers.get("plain") or "").strip():
            raise SnapshotError(f"fact {key!r} needs answers.plain")
        if "audio" in f:
            raise SnapshotError(f"fact {key!r}: Fact.audio forbidden in snapshot")
        disc = f.get("disclosure", "volunteered")
        if disc not in DISCLOSURES:  # unknown string → 400, never 500 in open (spec O)
            raise SnapshotError(f"fact {key!r}: bad disclosure {disc!r}")
        if key in keys:
            raise SnapshotError(f"duplicate fact {key!r}")
        keys.add(key)
    for c in raw.get("critical") or ():
        if c not in keys:
            raise SnapshotError(f"critical {c!r} not a fact")
    for f in facts:
        key = f.get("key") or f.get("id") or ""
        for r in f.get("requires") or ():
            if r not in keys:
                raise SnapshotError(f"fact {key!r} requires unknown {r!r}")
    return {
        "id": sid,
        "opening": raw["opening"],
        "critical": list(raw.get("critical") or ()),
        "facts": [
            {
                "key": f.get("key") or f.get("id"),
                "slot": f["slot"],
                "answers": dict(f["answers"]),
                "numbers": list(f.get("numbers") or ()),
                "requires": list(f.get("requires") or ()),
                "disclosure": f.get("disclosure", "volunteered"),
            }
            for f in facts
        ],
        "mode": mode,
    }
